lib/devfs: Add Doxygen-style comments

calex257 opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [ ] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [ ] Tested your changes against relevant architectures and platforms;
 - [ ] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

razvand commented:
@calex257, I would update this PR to be named `lib/devfs: Add Doxygen-style comments`. To encompass all files that you would update.

calex257 commented:
I should mention that the comments themselves are something I still have to work on, I hope that the format at least is close to what is desired.

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
 }
 
+/*
+ * Open the device specified by name
```
```suggestion
 * Opens the device specified by name.
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * Can be used to check whether the device exists or not
+ * 
+ * @param name
+ * The name of the device
```
```suggestion
 *     The name of the device
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param name
+ * The name of the device
+ * @param mode
+ * The open mode(should be handled by the device driver if needed)
```
```suggestion
 *     The open mode(should be handled by the device driver if needed)
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param mode
+ * The open mode(should be handled by the device driver if needed)
+ * @param devp
+ * Return parameter. The opened device will be reachable in devp after
```
```suggestion
 *     Return parameter. The opened device will be reachable in devp after
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * The open mode(should be handled by the device driver if needed)
+ * @param devp
+ * Return parameter. The opened device will be reachable in devp after
+ * the function returns
```
```suggestion
 *     the function returns.
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * Return parameter. The opened device will be reachable in devp after
+ * the function returns
+ * @return
+ * Returns the result of the device's open routine
```
```suggestion
 *     The result of the device's open routine
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
 int device_open(const char *name, int mode, struct device **devp);
+
+/*
+ * Close the device specified by dev
```
```suggestion
 * Closes the device specified by dev.
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+
+/*
+ * Close the device specified by dev
+ * Does not return error if the target driver does not have
```
```suggestion
 * The function does not return an error if the target driver does not have
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+/*
+ * Close the device specified by dev
+ * Does not return error if the target driver does not have
+ * a close routine
```
```suggestion
 * a close routine.
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * a close routine
+ * 
+ * @param dev
+ * The device to be closed
```
```suggestion
 *     The device to be closed
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * The structure containing the scatter-gather list
+ * @param ioflags
+ * The flags to be sent to the read operation
+ * @return
```
I think we should specify the exact result of the call, not that this function calls another one and returns its result. Similar to what @eduardvintila and I did in our PRs.

@razvand @StefanJum what do you think?

razvand replied:
Agreed. @calex257, please follow @RaduNichita's way of doing that.

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+/*
+ * Perform a scatter/gather read on the device pointed to by dev
+ * @param dev
+ * The device to read from
```
```suggestion
 *     The device to read from
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param dev
+ * The device to read from
+ * @param uio 
+ * The structure containing the scatter-gather list
```
```suggestion
 *     The structure containing the scatter-gather list
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
 int device_read(struct device *dev, struct uio *uio, int ioflags);
+
+/*
+ * Perform a scatter/gather write on the device pointed to by dev
```
```suggestion
 * Performs a scatter/gather write on the device pointed to by dev
 
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+/*
+ * Perform a scatter/gather write on the device pointed to by dev
+ * @param dev
+ * The device to write to
```
```suggestion
 *     The device to write to
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param dev
+ * The device to write to
+ * @param uio
+ * The structure containing the scatter-gather list
```
```suggestion
 *     The structure containing the scatter-gather list
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param uio
+ * The structure containing the scatter-gather list
+ * @param ioflags
+ * The flags to be sent to the write operation
```
```suggestion
 *     The flags to be sent to the write operation
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param ioflags
+ * The flags to be sent to the write operation
+ * @return
+ * The result of the write operation
```
```suggestion
 *     The result of the write operation
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
 int device_write(struct device *dev, struct uio *uio, int ioflags);
+
+/*
+ * Perform an I/O control request
```
```suggestion
 * Performs an I/O control request.
 
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+/*
+ * Perform an I/O control request
+ * @param dev
+ * The device to perform the request on
```
```suggestion
 *     The device to perform the request on
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param dev
+ * The device to perform the request on
+ * @param cmd
+ * The device dependent command to be executed
```
```suggestion
 *     The device dependent command to be executed
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param cmd
+ * The device dependent command to be executed
+ * @param arg
+ * The user buffer sent as argument to the command. The buffer
```
```suggestion
 *     The user buffer sent as argument to the command. The buffer
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * The device dependent command to be executed
+ * @param arg
+ * The user buffer sent as argument to the command. The buffer
+ * must be validated by the ioctl routine of the driver
```
```suggestion
 *     must be validated by the ioctl routine of the driver.
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * The user buffer sent as argument to the command. The buffer
+ * must be validated by the ioctl routine of the driver
+ * @return
+ * The result of the I/O control request
```
```suggestion
 *     The result of the I/O control request
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
 int device_ioctl(struct device *dev, unsigned long cmd, void *arg);
+
+/*
+ * Returns information about a device
```
```suggestion
 * Returns information about a device.
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+/*
+ * Returns information about a device
+ * The target of the query should be placed in the cookie field
+ * of the parameter
```
```suggestion
 * of the parameter
 
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * in the cookie field and on return the id, flags and name
+ * fields will be filled with the requested information.
+ * @return
+ * Returns 0 on success and ESRCH otherwise
```
```suggestion
 *     Used as a return parameter. The device must be specified
 *     in the cookie field and on return the id, flags and name
 *     fields will be filled with the requested information.
 * @return
 *     Returns 0 on success and ESRCH otherwise
```

@calex257 , check also the discussion [here](https://github.com/unikraft/unikraft/pull/708/files)

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
 int	nullop(void);
 
+/*
+ * Creates a new device object
```
```suggestion
 * Creates a new device object.
 
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
 int	enodev(void);
+
+/*
+ * Returns 0
```
```suggestion
 * Always returns 0.
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
 int bdev_write(struct device *dev, struct uio *uio, int ioflags);
 
+/*
+ * Returns ENODEV
```
```suggestion
 * Always returns ENODEV.
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+/*
+ * Creates a new device object
+ * @param drv
+ * The driver to create the device
```
```suggestion
 *     The driver to create the device
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param drv
+ * The driver to create the device
+ * @param name
+ * The name of the new device
```
```suggestion
 *     The name of the new device
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param name
+ * The name of the new device
+ * @param flags
+ * The flags used for the creation of the device
```
```suggestion
 *     The flags used for the creation of the device
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param flags
+ * The flags used for the creation of the device
+ * @return
+ * Returns the ID of the device on success and 0 otherwise
```
```suggestion
 *     Returns the ID of the device on success and 0 otherwise
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+
 struct device *device_create(struct driver *drv, const char *name, int flags);
+/*
+ * Destroy a device
```
```suggestion
 * Destroys a device.
 
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+/*
+ * Destroy a device
+ * @param dev
+ * The device to be destroyed
```
```suggestion
 *     The device to be destroyed
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param dev
+ * The device to be destroyed
+ * @return
+ * Returns 0 on success and ENODEV otherwise
```
```suggestion
 *     Returns 0 on success and ENODEV otherwise
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
 int device_destroy(struct device *dev);
+
+/*
+ * Destroys a locked device
```
```suggestion
 * Destroys a locked device.
 
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+
+/*
+ * Destroys a locked device
+ * Is equivalent to device_destroy but operates on locked devices
```
```suggestion
 * Similar to device_destroy() but operates on locked devices.
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * Destroys a locked device
+ * Is equivalent to device_destroy but operates on locked devices
+ * @param dev
+ * The device to be destroyed
```
```suggestion
 *     The device to be destroyed
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param dev
+ * The device to be destroyed
+ * @return
+ * Returns 0 on success and ENODEV otherwise
```
```suggestion
 *     Returns 0 on success and ENODEV otherwise
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
 int device_destroy_locked(struct device *dev);
+
+/*
+ * Register a device
```
```suggestion
 * Registers a device and prepends it to the list of devices.
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+
+/*
+ * Register a device
+ * Prepends the device to the list of devices.
```
```suggestion
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * The device is initialized as being active and having a reference count
+ * of 1.
+ * @param device
+ * The device to register
```
```suggestion
 *     The device to register
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param device
+ * The device to register
+ * @param name
+ * The name of the new device
```
```suggestion
 *     The name of the new device
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
+ * @param name
+ * The name of the new device
+ * @param flags
+ * The flags to be passed to the new device
```
```suggestion
 *     The flags to be passed to the new device
```

RaduNichita reviewed (changes requested):
The first round of review is done, @calex257.  

Most of them are formatting issues. Please apply the tabs from your editor, I think it may not be a good idea to apply the tab suggestion from the GUI interface.

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
 
+/*
+ * Open the device specified by name
+ * Can be used to check whether the device exists or not
```
```suggestion
 * This may be used to check whether a device exists or not.
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,151 @@ devtoname(struct device *dev)
 int device_close(struct device *dev);
+
+/*
+ * Perform a scatter/gather read on the device pointed to by dev
```
```suggestion
 * Perform a scatter/gather read on the device pointed to by dev.
 
```

razvand commented:
@cales257, please use the `Resolve conversation` button to remove those pieces of feedback that were taken care of.

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -48,19 +48,28 @@ struct device;
+	/* Device characteristics flags */
+	int		flags;
+	/* Device name */
+	char		name[MAXDEVNAME];
```
```suggestion
	/* Index cookie */
	unsigned long cookie;
	/* Device id */
	struct device *id;
	/* Device characteristics flags */
	int flags;
	/* Device name */
	char name[MAXDEVNAME];
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
+ * @param name
+ * 	The name of the device
+ * @param mode
+ * 	The open mode(should be handled by the device driver if needed)
```
```suggestion
 * 	The open mode (should be handled by the device driver if needed)
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
+ * 	The flags to be passed to the new device
+ */
 void device_register(struct device *device, const char *name, int flags);
 void read_partition_table(struct device *device);
```
Please also comment this function.

calex257 replied:
I have searched in the repository and haven't found any defnition for it. I think the [older version](https://github.com/cloudius-systems/osv/commit/7fb8b99b3cc8239eb990949ccd958d5a243aa424) has been removed in this [commit](https://github.com/unikraft/unikraft/commit/b41122a00c0ca3f4fa8fba26701e929efc345419#diff-f5d57dea8605432735b89cff10b13badac448f0bc78edf8253f761539dba4979).

razvand replied:
Remove the declaration then.

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
+/*
+ * Registers a device and prepends the device to the list of devices.
+ * If name is longer than MAXDEVNAME characters returns without creating the device.
+ * If the device is already registered the function triggers an error.
```
> If the device is already registered the function triggers an error

It is not clear to me where is the error triggered? Could you be more explicit, please?

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
+ * 	The device to register
+ * @param name
+ * 	The name of the new device
+ * @param flags
```
Could you add some of the most used options for flags?

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
+ * @return
+ * 	The ID of the device on success and 0 otherwise
+ */
+
```
```suggestion
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
+ * @param flags
+ * 	The flags used for the creation of the device
+ * @return
+ * 	The ID of the device on success and 0 otherwise
```
What is an ID? I think a structure is returned from this function.

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
+/*
+ * Always returns ENODEV.
+ */
 int	enodev(void);
```
```suggestion
int enodev(void);
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
+/*
+ * Always returns 0.
+ */
 int	nullop(void);
```
```suggestion
int nullop(void);
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
+
 int device_info(struct devinfo *info);
 
 int bdev_read(struct device *dev, struct uio *uio, int ioflags);
```
Please also comment this function.

calex257 replied:
Same as with `read_partition_table`, I could not find any definition or reference to it in the repository. The only reference I could find for them is [here](https://github.com/cloudius-systems/osv/blob/17bb6ae35c31b79615913a4a0363dae0a151c6f6/fs/vfs/vfs_bdev.cc).

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
 int device_info(struct devinfo *info);
 
 int bdev_read(struct device *dev, struct uio *uio, int ioflags);
 int bdev_write(struct device *dev, struct uio *uio, int ioflags);
```
Please also comment this function.

calex257 replied:
Same as for `bdev_read`.

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -95,42 +104,62 @@ struct devops {
-	struct devops	*devops;	/* device operations */
-	size_t		devsz;		/* size of private data */
-	int		flags;		/* state of driver */
+	/* name of device driver */
+	const char	*name;
+	/* device operations */
+	struct devops	*devops;
+	/* size of private data */
+	size_t		devsz;
+	/* state of driver */
+	int		flags;
```
```suggestion
	/* name of device driver */
	const char *name;
	/* device operations */
	struct devops *devops;
	/* size of private data */
	size_t devsz;
	/* state of driver */
	int flags;
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
+ * @return
+ * 	The result of the device's close routine
+ */
+
```
```suggestion
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
+ * @return
+ * 	The result of the device's open routine
+ */
+
```
```suggestion
```

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,158 @@ devtoname(struct device *dev)
+ * @return
+ * 	0 on success and ESRCH otherwise
+ */
+
```
```suggestion
```

RaduNichita reviewed (changes requested):
Another round. Will do a deeper pass in the next days.

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,157 @@ devtoname(struct device *dev)
+ * @param mode
+ * 	The open mode (should be handled by the device driver if needed)
+ * @param devp
+ * 	Return parameter. The opened device will be reachable in devp after
```
This should be marked as an output parameter. See the example here [1]

[1] https://github.com/unikraft/unikraft/blob/staging/lib/vfscore/vfs.h#L277

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -95,42 +104,62 @@ struct devops {
+	char name[MAXDEVNAME];
+	/* D_* flags defined above */
+	int flags;
+	/* Device has not been destroyed */
```
Maybe we can change this to status of the device? And mention that 1 means active and 0 inactive. What do you think?

calex257 replied:
Yeah, it's a clearer description.

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,157 @@ devtoname(struct device *dev)
+ * @param dev
+ * 	The device to be destroyed
+ * @return
+ * 	0 on success and ENODEV otherwise
```
Please use the coding convention using lists with bullets. You can see an example at [1]

[1] https://github.com/unikraft/unikraft/blob/staging/lib/vfscore/vfs.h#L304

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -190,22 +219,157 @@ devtoname(struct device *dev)
+ * of the parameter.
+ *
+ * @param info
+ * 	Used as a return parameter. The device must be specified
```
Mark this as an output parameter, as suggested in another comment.

RaduNichita commented on lib/devfs/include/devfs/device.h:
```diff
@@ -95,42 +104,62 @@ struct devops {
+	/* Private storage */
+	void *private_data;
 
 	void *softc;
```
Please comment on these members of the structure.

RaduNichita reviewed (changes requested):
Another round.

StefanJum commented:
@calex257 is this pr ready for review? If so please mark it as not draft.
Also please mark the comments as resolved after you fix them.

razvand reviewed (commented):
@calex257, see my comments and let's ready this PR for the 0.13 release.

razvand commented:
@calex257, please solve the conflicts. And also coalesce all commits into a single one.

calex257 committed:
lib/devfs: Add Doxygen-style comments

Add Doxygen-style comments to `indelude/devfs/device.h`, the header file
where `devfs` structures and functions are defined.

Signed-off-by: Alexandru Calciu <alexandru.calciu@stud.acs.upb.ro>

StefanJum commented on lib/devfs/include/devfs/device.h:
```diff
@@ -86,44 +98,179 @@ struct devops {
+ */
 int device_create(struct driver *drv, const char *name, int flags,
 		  struct device **devp);
+/**
```
```suggestion

/**
```

StefanJum commented on lib/devfs/include/devfs/device.h:
```diff
@@ -86,44 +98,179 @@ struct devops {
+ *
+ * @param fn
+ * 	The initilization function to be called
+ * 	Sample functions: devfs_register, devfs_register_null,
+ * 	  devfs_register_stdout
```
```suggestion
 * 	Sample functions: `devfs_register()`, `devfs_register_null()`,
 * 	  `devfs_register_stdout()`.
```

StefanJum reviewed (changes requested):
Few inline comments.
Besides those, please do not use tabs inside comment blocks, since they mess up the doxygen renderer. Use 4 spaces instead.

razvand commented:
> Few inline comments. Besides those, please do not use tabs inside comment blocks, since they mess up the doxygen renderer. Use 4 spaces instead.

Thanks, @StefanJum , I fixed the comments.

StefanJum reviewed (approved):
All good.
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
